### PR TITLE
Overrides for AbstractSimpleGraph

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
The docs don't explicitly specify if all AbstractGraph types should return neighbors in sorted order.But assuming it to be true (true for SimpleGraph atleast) , the speed of common_neighbors can be improved.

Here are some benchmarks for reference - 

 ```julia
julia> g = CompleteGraph(1000)
{1000, 499500} undirected simple Int64 graph

julia> @benchmark common_neighbors(g, rand(1:1000), rand(1:1000))

BenchmarkTools.Trial: 
  memory estimate:  16.45 KiB
  allocs estimate:  10
  --------------
  minimum time:     15.006 μs (0.00% GC)
  median time:      16.695 μs (0.00% GC)
  mean time:        25.341 μs (25.92% GC)
  maximum time:     61.196 ms (99.85% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> old_common_neighbors(g::AbstractGraph, u::Integer, v::Integer) =
       intersect(neighbors(g, u), neighbors(g, v))

old_common_neighbors (generic function with 1 method)

julia> @benchmark old_common_neighbors(g, rand(1:1000), rand(1:1000))

BenchmarkTools.Trial: 
  memory estimate:  108.44 KiB
  allocs estimate:  38
  --------------
  minimum time:     145.887 μs (0.00% GC)
  median time:      150.034 μs (0.00% GC)
  mean time:        166.843 μs (6.93% GC)
  maximum time:     61.436 ms (99.58% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

### CODE REVIEW

**Comment by sbromberger** [comment]
I'm not sure that it's part of the API guarantee, but we could make it one. It might break third-party graph implementations but we don't currently have an API guarantee on `neighbors` yet anyway.

**Comment by SohamTamba** [comment]
Is it possible to keep both implementations?

```
function common_neighbors(g::SimpleGraph, u::Integer, v::Integer)
    THIS IMPLEMENTATION
end
````
`common_neighbors(g::NotSimpleGraph, u::Integer, v::Integer) = intersect(neighbors(g, u), neighbors(g, v))`

**Comment by abhinavmehndiratta** [comment]
> Is it possible to keep both implementations?
> 
> ```
> function common_neighbors(g::SimpleGraph, u::Integer, v::Integer)
>     THIS IMPLEMENTATION
> end
> ```
> 
> `common_neighbors(g::NotSimpleGraph, u::Integer, v::Integer) = intersect(neighbors(g, u), neighbors(g, v))`
> 
> Another option is make change the type of `adjlist` to `SortedVector` and let multiple dispatch take over.
> The runtime for `intersection` of `SortedVector`, `Vector` is linear, quadratic respectively.
> 
> The second option might be better because similar optimisations can be made while capturing the assumption that `SimpleGraph` has a sorted adjacency list.

The second approach sounds better but I am not sure if SortedVector is part of Base ? The other way is also possible by overloading the method for simple graphs.

**Comment by SohamTamba** [comment]
> > Is it possible to keep both implementations?
> > ```
> > function common_neighbors(g::SimpleGraph, u::Integer, v::Integer)
> >     THIS IMPLEMENTATION
> > end
> > ```
> > `common_neighbors(g::NotSimpleGraph, u::Integer, v::Integer) = intersect(neighbors(g, u), neighbors(g, v))`
> > Another option is make change the type of `adjlist` to `SortedVector` and let multiple dispatch take over.
> > The runtime for `intersection` of `SortedVector`, `Vector` is linear, quadratic respectively.
> > The second option might be better because similar optimisations can be made while capturing the assumption that `SimpleGraph` has a sorted adjacency list.
> 
> The second approach sounds better but I am not sure if SortedVector is part of Base ? The other way is also possible by overloading the method for simple graphs.

They have [Sorted Set](https://github.com/JuliaCollections/DataStructures.jl/blob/master/src/sorted_set.jl) but that is implemented using a tree, not an array (Let's not use that)

**Comment by sbromberger** [comment]
I think we should just make the API guarantee that `neighbors` returns in sorted order, and then any graph type that doesn't do that will need to change `neighbors` and/or override this for its own particular performance needs.

cc @jpfairbanks for a discussion on the API proposal.

**Comment by SohamTamba** [comment]
Then we might have to mark in the documentation which functions make the sorted assumptions.

I wouldn't want someone to think they have to rewrite Blossom V (when they don't) because they are using a graph represented by an unsorted adjacency list.

**Comment by abhinavmehndiratta** [comment]
I we require that the vertices are in range 1:nv(g) then it should not be too expensive to require that the neighbors are sorted (using radix sort).

File: src/core.jl [hunk 1]
```diff
@@ -314,8 +311,30 @@ julia> common_neighbors(g, 1, 4)
...
+    common = Vector{Int}()
+    u_nbrs, v_nbrs = neighbors(g, u), neighbors(g, v)
+    i, j = 1, 1
+    u_len, v_len = length(u_nbrs), length(v_nbrs)
+
+    if (u_nbrs[u_len] < v_nbrs[1]) || (u_nbrs[1] > v_nbrs[v_len])
+       return common
+    end
+
+    while i <= u_len && j <= v_len
```
Comment (simonschoelly) [inline src/core.jl]: Prefix loop with `@inbounds`

File: src/core.jl [hunk 2]
```diff
@@ -314,8 +311,30 @@ julia> common_neighbors(g, 1, 4)
  3
 ```
 """
-common_neighbors(g::AbstractGraph, u::Integer, v::Integer) =
-    intersect(neighbors(g, u), neighbors(g, v))
+function common_neighbors(g::AbstractGraph, u::Integer, v::Integer)
+    common = Vector{Int}()
```
Comment (simonschoelly) [inline src/core.jl]: Should be `Vector{eltype(g)}`. Can you please also add a test, that verifies that the result has the same eltype as the input graph?

File: src/core.jl [hunk 3]
```diff
@@ -314,8 +311,30 @@ julia> common_neighbors(g, 1, 4)
...
+    u_nbrs, v_nbrs = neighbors(g, u), neighbors(g, v)
+    i, j = 1, 1
+    u_len, v_len = length(u_nbrs), length(v_nbrs)
+
+    if (u_nbrs[u_len] < v_nbrs[1]) || (u_nbrs[1] > v_nbrs[v_len])
```
Comment (simonschoelly) [inline src/core.jl]: If you write `u_nbrs[end]` instead of `u_nbrs[u_len]`, the intention might be a bit easier to see.

**Comment by simonschoelly** [comment]
I don't think the api should guarantee sorted order for all vertices (at least not until we are sure that we don't have another graph type with unsorted vertices in the future). 
I think it might be easier to either restrict this method to `AbstractSimpleGraph` or test for sorted order in the method and then do something different if they are not sorted.

File: src/core.jl [hunk 4]
```diff
@@ -314,8 +311,30 @@ julia> common_neighbors(g, 1, 4)
...
 """
-common_neighbors(g::AbstractGraph, u::Integer, v::Integer) =
-    intersect(neighbors(g, u), neighbors(g, v))
+function common_neighbors(g::AbstractGraph, u::Integer, v::Integer)
+    common = Vector{eltype(g)}()
+    u_nbrs, v_nbrs = neighbors(g, u), neighbors(g, v)
+    i, j = 1, 1
+    u_len, v_len = length(u_nbrs), length(v_nbrs)
+
+    if (u_nbrs[end] < v_nbrs[1]) || (u_nbrs[1] > v_nbrs[end])
```
Comment (simonschoelly) [inline src/core.jl]: I think this will be out of bounds, when a vertex has 0 neighbors.

Comment (abhinavmehndiratta) [reply src/core.jl]: Fixed it, I don't know why I miss the corner cases :man_facepalming:

Comment (simonschoelly) [reply src/core.jl]: I could place to think about what corner cases are, is when you are writing tests.

**Comment by simonschoelly** [comment]
I think this is basically good to go, but maybe the issue with using which graph type should be decided first.

On a related note, this function also works for directed graphs, where (in LightGraphs) `neighbors` is the as `outneighbors`. Is that also the behavior that we want here and should that be documented?

**Comment by sbromberger** [comment]
Let's make this `AbstractSimpleGraph` for right now, since it's easy to provide the guarantee that the neighbors are in sorted order.

Once we do that, we should move it into `SimpleGraphs/` directory, maybe in `overrides.jl`?

**Comment by abhinavmehndiratta** [comment]
> Let's make this `AbstractSimpleGraph` for right now, since it's easy to provide the guarantee that the neighbors are in sorted order.
> 
> Once we do that, we should move it into `SimpleGraphs/` directory, maybe in `overrides.jl`?

Okay, but is there a graph type other than AbstractSimpleGraph so that we can add a test that the core version of the function works for other types as well ? It might go untested if we overload it for SimpleGraphs.

**Comment by abhinavmehndiratta** [comment]
Should the tests be in `test/core.jl` or `test/simplegraphs/simplegraphs.jl` ?

**Comment by simonschoelly** [comment]
I think the test should be in test/simplegraphs/overrides.jl, so that we know where to find it.
About testing with a non `AbstractSimpleGraph`, I can think of 3 options:

1. Use SimpleWeightedGraphs. These graphs are not subtypes of `AbstractSimpleGraph`. This might change in the future though. Therefore, if you do this, you should add a warning, that gets triggered when this graph is suddenly a subtype of `AbstractSimpleGraph`.

2. You can use the function `invoke` to trigger the function for the abstract type instead. Unfortunately I haven't figured out, how to use it, when your function uses `@traitfn`.

3. Create a special graph just for testing this. This could maybe be a wrapper around a `SimpleGraph`, that forwards all function calls to the wrapped graph. This might be a bit of work though. 

Btw. I had a similar problem in #1064 which was never merged.

**Comment by sbromberger** [comment]
> I think the test should be in test/simplegraphs/overrides.jl, so that we know where to find it.

I'm not opposed to this but I think the convention for the core (non-interface) functions is to put the "overrides" in the main .jl file. (In this case, `SimpleGraphs.jl`.)

**Comment by simonschoelly** [comment]
That makes also sense

**Comment by abhinavmehndiratta** [comment]
I am using `invoke` to explicitly call the core function and moved the tests to `test/simplegraphs/simplegraphs.jl`

**Comment by abhinavmehndiratta** [comment]
For some reason `@inferred` does not work with invoke, so I've removed it for now.
```julia
julia> function f(x)
           return x+1
       end
f (generic function with 1 method)

julia> invoke(f, Tuple{Int}, 1)
2

julia> @inferred(invoke(f, Tuple{Int}, 1))
ERROR: ArgumentError: argument is not a generic function
Stacktrace:
 [1] return_types(::Any, ::Any) at ./reflection.jl:892
 [2] top-level scope at none:0
```

**Comment by abhinavmehndiratta** [comment]
Benchmarks for intersect 
```julia
julia> g1 = loadsnap(:ego_twitter_d)
{81306, 1768149} directed simple Int64 graph

julia> g2 = loadsnap(:soc_slashdot0902_d)
{82168, 948464} directed simple Int64 graph

julia> @btime intersect(g1, g2) #old
  2.072 s (10283376 allocations: 475.62 MiB)
{81306, 1782} directed simple Int64 graph

julia> @btime intersect(g1, g2) #new
  29.439 ms (165117 allocations: 13.77 MiB)
{81306, 1782} directed simple Int64 graph
```

**Comment by simonschoelly** [comment]
That is a lot faster.

Let's not add more different overrides to this same pr though.

I've been thinking, that testing with inferred might be dangerous, as when another function is used inside a function, then `invoke` might not work as expected. Here is an example
```julia
julia> abstract type AT end
julia> struct T <: AT end
julia> f(::AT) = "AT"
julia> f(::T) = "T"
julia> g(x) = f(x)
julia> invoke
julia> invoke(f, Tuple{AT}, T())
"AT"
julia> invoke(f, Tuple{T}, T())
"T"
julia> invoke(g, Tuple{AT}, T())
"T"
```

So I think the best thing is to create test structures `DummyAbstractGraph`  or `AbstractTestGraph` or something else (and the same for `DiGraph`)  that is basically just a wrapper around a simple graph, but is not a subtype of any other type than `AbstractGraph`.  Using macros, it should be easy to create the LightGraph interface functions for this graph, that then call the function on the wrapped type.

**Comment by abhinavmehndiratta** [comment]
> So I think the best thing is to create test structures `DummyAbstractGraph` or `AbstractTestGraph` or something else (and the same for `DiGraph`) that is basically just a wrapper around a simple graph, but is not a subtype of any other type than `AbstractGraph`. Using macros, it should be easy to create the LightGraph interface functions for this graph, that then call the function on the wrapped type.

@simonschoelly agreed, although I don't understand why `g(x) = f(x)` only creates g with 1 method instead of 2. @sbromberger if I create a dummy class in which directory should it go ?

**Comment by sbromberger** [comment]
> @sbromberger if I create a dummy graph structure in which directory should it go ?

This is for tests, right? It should go into the local test file. We have examples of these throughout.

**Comment by simonschoelly** [comment]
Yes, dummy graph is for testing only, so that when we overload an abstract method, that we can still test the abstract method.

File: src/SimpleGraphs/SimpleGraphs.jl [hunk 5]
```diff
@@ -33,8 +34,9 @@ export AbstractSimpleGraph, AbstractSimpleEdge,
     #smallgraphs
     smallgraph,
     # Euclidean graphs
-    euclidean_graph
-
+    euclidean_graph,
+    #overrides
+    common_neighbors, intersect
```
Comment (simonschoelly) [inline src/SimpleGraphs/SimpleGraphs.jl]: It is not necessary to export overwritten symbols.
We could still do it though, so it works as a documentation.

File: test/operators.jl [hunk 6]
```diff
@@ -1,3 +1,16 @@
+import LightGraphs.edges, LightGraphs.nv, LightGraphs.add_edge!
+
+abstract type DummyAbstractGraph{T} <: AbstractGraph{T} end
+
+mutable struct DummyGraph{T} <: DummyAbstractGraph{T}
+    SG::SimpleGraph{T}
+end
```
Comment (simonschoelly) [inline test/operators.jl]: We define `DummyAbstractGraph` in two different locations. Wouldn't it be better to have a `testutils.jl` (or similar name) file, where we define such things?

File: src/SimpleGraphs/SimpleGraphs.jl (hunk 5 above)
Comment (abhinavmehndiratta) [reply src/SimpleGraphs/SimpleGraphs.jl]: Yes, I did that so that it is easy to remember that the method is overridden, can remove it though.

File: test/operators.jl (hunk 6 above)
Comment (abhinavmehndiratta) [reply test/operators.jl]: Yes the build is also failing due to redefinition of DummyGraph (should have used a different name) but creating testutils.jl is a better idea

Comment (abhinavmehndiratta) [reply test/operators.jl]: But that should be part of a separate PR I guess, @sbromberger views ?

Comment (matbesancon) [reply test/operators.jl]: @abhinavmehndiratta you can just change the name here to `NewDummyGraph`, this would do it

Comment (abhinavmehndiratta) [reply test/operators.jl]: Okay I'll fix it for now :smile:

**Comment by abhinavmehndiratta** [comment]
The build failure is unrelated, please restart it.

**Comment by matbesancon** [comment]
Restarted. If you want to do it when no-one is around, you can close and re-open the PR

**Comment by abhinavmehndiratta** [comment]
> Restarted. If you want to do it when no-one is around, you can close and re-open the PR

Thank you ! Didn't realize that.
